### PR TITLE
Main api

### PR DESCRIPTION
**Pull request by toabctl**

### CODE REVIEW

File: awspub/api.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    images: List[Tuple[str, Image, Dict[str, _ImageInfo]]] = []
+    for image_name, image in _images_filtered(ctx, group):
+        image_result: Dict[str, _ImageInfo] = image.list()
+        images.append((image_name, image, image_result))
+
+    images_by_name, images_by_group = _images_grouped(images, group)
+    return images_by_name, images_by_group
+
+
+def public(config: pathlib.Path, config_mapping: pathlib.Path, group: Optional[str]):
```
Comment (rpocase) [inline awspub/api.py:114]: `make_public` would be a bit better. `public` doesn't really imply any sort of action imo

File: docs/reference/logic.rst [hunk 2]
```diff
@@ -1,5 +1,18 @@
-Reference for logical components
-================================
+Reference for the API
+=====================
+
+If in doubt, use the high level API. This
+should work well for 95% of the use cases.
```
Comment (rpocase) [inline docs/reference/logic.rst:5]:
```suggestion
should work well for most use cases.
```

**Review by rpocase — DISMISSED** [review]
overall lgtm. couple nits

File: awspub/api.py (hunk 1 above)
Comment (toabctl) [reply awspub/api.py:114]: yeah. I kind of like the single word but agree that `public` isn't good. `publish` would be another option. 
Any strong opinions?

Comment (rpocase) [reply awspub/api.py:114]: i'm happy with `publish`.

Comment (toabctl) [reply awspub/api.py:114]: for some reason (that I currently don't remember) I back then decided to not use `publish`. 
If we use `publish` now for the API, the CLI would still use `public`. Do we want that inconsistency here?

Comment (rpocase) [reply awspub/api.py:114]: i would change the CLI at the same time unless we can remember a compelling reason why publish didn't make sense in the cli api.

**Review by JessicaJang — DISMISSED** [review]

File: awspub/api.py (hunk 1 above)
Comment (JessicaJang) [reply awspub/api.py:114]: im towards more in public since `publish` serves meaning like whole operation. For representing meaning of this function, I would use `public`.

File: docs/reference/logic.rst (hunk 2 above)
Comment (toabctl) [reply docs/reference/logic.rst]: done

File: awspub/api.py (hunk 1 above)
Comment (toabctl) [reply awspub/api.py]: Ryan also said he prefers `publish`. So I added an extra commit to change the name for the API and the CLI.

File: awspub/api.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+from awspub.context import Context
+from awspub.image import Image, _ImageInfo
+from awspub.s3 import S3
+
+logger = logging.getLogger(__name__)
+
+
+def _images_grouped(
+    images: List[Tuple[str, Image, Dict[str, _ImageInfo]]], group: Optional[str]
+) -> Tuple[Dict[str, Dict[str, str]], Dict[str, Dict[str, Dict[str, str]]]]:
```
Comment (gjolly) [inline awspub/api.py:14]: You could define your custom types here. Like:

```python3
type ImageMetadata = Dict[str, str]
```

File: awspub/api.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    Group the given images by name and by group
+
+    :param images: the images
+    :type images: List[Tuple[str, Image, Dict[str, _ImageInfo]]]
+    :param group: a optional group name
+    :type group: Optional[str]
+    :return: the images grouped by name and by group
+    :rtype: Tuple[Dict[str, Dict[str, str]], Dict[str, Dict[str, Dict[str, str]]]
```
Comment (gjolly) [inline awspub/api.py:23]: Why is this needed, it's already in the signature above.

**Review by gjolly — APPROVED** [review]
`mypy` strict gives me:

```
awspub/api.py:114: error: Function is missing a return type annotation  [no-untyped-def]
awspub/api.py:131: error: Function is missing a return type annotation  [no-untyped-def]
awspub/api.py:148: error: Missing type parameters for generic type "Dict"  [type-arg]
awspub/api.py:161: error: Missing type parameters for generic type "Dict"  [type-arg]
Found 4 errors in 1 file (checked 1 source file)
```

But you haven't configured mypy with `--strict` in `tox.ini` so I don't know if you want to fix that.

File: awspub/api.py (hunk 4 above)
Comment (toabctl) [reply awspub/api.py:23]: looks like there is https://github.com/tox-dev/sphinx-autodoc-typehints that might help here.

File: awspub/api.py (hunk 3 above)
Comment (toabctl) [reply awspub/api.py:14]: `type` is >= python 3.12  (see https://docs.python.org/3/reference/simple_stmts.html#type ). but we want to support 3.10 (from jammy) for now.

**Comment by toabctl** [comment]
> `mypy` strict gives me:
> 
> ```
> awspub/api.py:114: error: Function is missing a return type annotation  [no-untyped-def]
> awspub/api.py:131: error: Function is missing a return type annotation  [no-untyped-def]
> awspub/api.py:148: error: Missing type parameters for generic type "Dict"  [type-arg]
> awspub/api.py:161: error: Missing type parameters for generic type "Dict"  [type-arg]
> Found 4 errors in 1 file (checked 1 source file)
> ```
> 
> But you haven't configured mypy with `--strict` in `tox.ini` so I don't know if you want to fix that.

Created https://github.com/canonical/awspub/issues/55 to enable `strict` mode.

File: awspub/api.py (hunk 4 above)
Comment (toabctl) [reply awspub/api.py:23]: Filled https://github.com/canonical/awspub/issues/56 to explore this more.

**Review by JessicaJang — APPROVED** [review]

**Review by rpocase — APPROVED** [review]
